File: grep-error/grep-error.c
#include <stdio.h>
#include <unistd.h>
#include <string.h>

int main() {
    int fd[2];
    pipe(fd);

    if (fork() == 0) {
        // Filho = filtro (grep ERROR)
        close(fd[1]);

        char buffer[256];
        ssize_t n;
        while ((n = read(fd[0], buffer, sizeof(buffer) - 1)) > 0) {
            buffer[n] = '\0';  
        
            if (strstr(buffer, "ERROR")) {
                printf("Filtrado: %s", buffer);
            }
            
        }
    } else {
        // Pai = produtor de logs
        close(fd[0]);

        char *logs[] = {
            "INFO Sistema OK\n",
            "ERROR Falha no disco\n",
            "INFO Reiniciando\n",
            "ERROR Timeout\n"
        };

        for (int i = 0; i < 4; i++) {
            write(fd[1], logs[i], strlen(logs[i]));
            sleep(1);
        }
    }
}

File: grep-error/grep-error.py
import time
from multiprocessing import Process, Pipe

def filtro(conn):
    while True:
        msg = conn.recv()
        
        if msg is None:
            break
        
        if "ERROR" in msg:
            print("Filtrado:", msg)
    
    conn.close()

parent, child = Pipe()

p = Process(target=filtro, args=(child,))
p.start()

logs = [
    "INFO Sistema OK",
    "ERROR Falha no disco",
    "INFO Reiniciando",
    "ERROR Timeout"
]

for log in logs:
    parent.send(log)
    time.sleep(1)

parent.send(None)
parent.close()
p.join()
